sys/kern/exception.c

```diff
@@ -17,18 +21,50 @@ void on_exc_leave(void) {
   }
 }
 
-void on_user_exc_leave(void) {
-  thread_t *td = thread_self();
+static void set_syscall_retval(user_ctx_t *ctx, syscall_result_t *result,
+                               int sig) {
+  int error = result->error;
+  proc_t *p = proc_self();
+
+  if (error == EJUSTRETURN)
+    return;
 
+  if (error == ERESTARTSYS || error == ERESTARTNOHAND) {
+    if (!sig || (error == ERESTARTSYS &&
+                 (p->p_sigactions[sig].sa_flags & SA_RESTART))) {
+      user_ctx_restart_syscall(ctx);
+      return;
+    }
+    error = EINTR;
+  }
+
+  user_ctx_set_retval(ctx, result->retval, error);
+}
+
+void on_user_exc_leave(user_ctx_t *ctx, syscall_result_t *result) {
+  thread_t *td = thread_self();
   proc_t *p = td->td_proc;
-  /* Process pending signals. */
+  int sig = 0;
+  ksiginfo_t ksi;
+
   if (td->td_flags & TDF_NEEDSIGCHK) {
+    WITH_PROC_LOCK(p) {
+      sig = sig_check(td, &ksi);
+    }
+  }
+
+  /* Set return value from syscall, restarting it if needed. */
+  if (result)
+    set_syscall_retval(ctx, result, sig);
+
+  /* Process pending signals. */
+  while (sig) {
     WITH_PROC_LOCK(p) {
       /* Calling sig_post() multiple times before returning to userspace
-       * will not make us lose signals, see comment on sig_post() in signal.h */
-      ksiginfo_t ksi;
-      while (sig_check(td, &ksi))
-        sig_post(&ksi);
+       * will not make us lose signals, see comment on sig_post() in signal.h
+       */
+      sig_post(&ksi);
+      sig = sig_check(td, &ksi);
```

Comment:
I don't understand. What is the difference?

Reply:
This is a bit ugly. We need to know whether a signal will be caught in order to set the syscall return value (and maybe restart the syscall), so we must call `sig_check()` before `set_syscall_retval()`.  But we also must call `set_syscall_retval()` before `sig_post()`, as `sig_post()` modifies the user context.

Reply:
Thanks for your explanation.